### ISSUE TITLE
Implement Mutex

### ISSUE THREAD

**Issue by lopopolo** [artichoke/artichoke#291]
Implement `Mutex`: https://ruby-doc.org/core-2.6.3/Mutex.html

Ruby mutexes are not reentrant, so we can use the Mutex implementation in std, maybe with [`parking_lot`](https://github.com/Amanieu/parking_lot) as an optional dependency.

The API is small and likely needs to be implemented entirely in Rust with the exception of `Mutex#synchronize`.

There is an existing single-threaded implementation of `Mutex` that we need to remove.

https://github.com/artichoke/artichoke/blob/9f40627dbd2725d97d6b8ac7774993efa5fdd87a/artichoke-backend/src/extn/core/thread.rb#L297-L344

Unblocks GH-253.

**Comment by Madao-3** [comment]
@lopopolo  I can work on this.

**Comment by lopopolo** [comment]
Thanks @Madao-3 

One thing to keep in mind is that mruby is not threadsafe and without `Thread` this won’t be that useful. This work will provide the groundwork.

One way to make it useful is to allow the mutex to be shared across multiple interpreters, but that can be addressed in a separate ticket if there is a use case for it.
